sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala

```diff
@@ -121,43 +121,81 @@ abstract class ParquetFilterSuite extends QueryTest with ParquetTest with Shared
     checkBinaryFilterPredicate(predicate, filterClass, Seq(Row(expected)))(df)
   }
 
+  /**
+   * Takes single level `inputDF` dataframe to generate multi-level nested
+   * dataframes as new test data.
+   */
+  private def withNestedDataFrame(inputDF: DataFrame)
```

Comment:
how about
```
def withNestedDataFrame[T <: Product: ClassTag: TypeTag](data: Seq[T])... {
  val df = data.toDF("temp")
}
```

Otherwise, the `inputDF` can be any schema and this method looks fragile.

Reply:
Sounds good.

I think ` private def withNestedDataFrame[T <: Product: ClassTag: TypeTag](data: Seq[T])` is good enough. I feel no reason to put it in `Tuple1`.

Reply:
Okay, this is not easy since one of the test case is like
```scala
        val dataFrame = spark.createDataFrame(rdd, StructType.fromDDL(s"a decimal($precision, 2)"))
        withNestedDataFrame(dataFrame) { case (inputDF, pushDownColName, resultFun) =>
          withParquetDataFrame(inputDF) { implicit df =>
            val decimalAttr: Expression = df(pushDownColName).expr
            assert(df(pushDownColName).expr.dataType === DecimalType(precision, 2))
```
, and the dataframe can not be constructed directly from `withNestedDataFrame[T](data: Seq[T])`

Reply:
Instead, I add schema checking in the code.